src/librustc_middle/ty/mod.rs

```diff
@@ -1347,86 +1376,114 @@ impl ToPredicate<'tcx> for PredicateKind<'tcx> {
     }
 }
 
-impl<'tcx> ToPredicate<'tcx> for ConstnessAnd<TraitRef<'tcx>> {
+impl ToPredicate<'tcx> for PredicateAtom<'tcx> {
+    #[inline(always)]
     fn to_predicate(&self, tcx: TyCtxt<'tcx>) -> Predicate<'tcx> {
-        ty::PredicateKind::Trait(
-            ty::Binder::dummy(ty::TraitPredicate { trait_ref: self.value }),
-            self.constness,
-        )
-        .to_predicate(tcx)
+        tcx.mk_predicate(ty::PredicateKind::Atom(*self))
     }
 }
 
-impl<'tcx> ToPredicate<'tcx> for ConstnessAnd<&TraitRef<'tcx>> {
+impl<'tcx> ToPredicate<'tcx> for ConstnessAnd<TraitRef<'tcx>> {
     fn to_predicate(&self, tcx: TyCtxt<'tcx>) -> Predicate<'tcx> {
-        ty::PredicateKind::Trait(
-            ty::Binder::dummy(ty::TraitPredicate { trait_ref: *self.value }),
-            self.constness,
-        )
-        .to_predicate(tcx)
+        ty::PredicateAtom::Trait(ty::TraitPredicate { trait_ref: self.value }, self.constness)
+            .to_predicate(tcx)
     }
 }
 
 impl<'tcx> ToPredicate<'tcx> for ConstnessAnd<PolyTraitRef<'tcx>> {
     fn to_predicate(&self, tcx: TyCtxt<'tcx>) -> Predicate<'tcx> {
-        ty::PredicateKind::Trait(self.value.to_poly_trait_predicate(), self.constness)
-            .to_predicate(tcx)
+        ConstnessAnd {
+            value: self.value.map_bound(|trait_ref| ty::TraitPredicate { trait_ref }),
+            constness: self.constness,
+        }
+        .to_predicate(tcx)
     }
 }
 
-impl<'tcx> ToPredicate<'tcx> for ConstnessAnd<&PolyTraitRef<'tcx>> {
+impl<'tcx> ToPredicate<'tcx> for ConstnessAnd<PolyTraitPredicate<'tcx>> {
     fn to_predicate(&self, tcx: TyCtxt<'tcx>) -> Predicate<'tcx> {
-        ty::PredicateKind::Trait(self.value.to_poly_trait_predicate(), self.constness)
+        if let Some(pred) = self.value.no_bound_vars() {
```

Comment:
I wonder if this could be expressed as something like

```rust
self.value.map_bound(|pred| ty::PredicateAtom::Trait(pred, self.constness)).to_predicate(tcx)
```

where we have an impl for `Binder<PredicateAtom<'tcx>>` that either drops the binder (if nothing is escaping) or else wraps in `ForAll`?

Would encapsulate this pattern I see in a few places, I think.

Reply:
I don't want to implicitly wrap `Predicate`s with forall, which is why I call `potentially_quantified` with `potentially_quantified(tcx, PredicateKind::ForAll)` not sure how much this helps, but I think that if we ever add another quantifier this should make it easier to prevent mistakes here.

Changed this to the following for now, which seems fairly nice to me:
```rust
PredicateAtom::Trait(self.value.skip_binder(), self.constness)
            .potentially_quantified(tcx, PredicateKind::ForAll)
            .to_predicate(tcx)
```